python/ray/tune/tests/test_syncer.py

```diff
@@ -440,6 +442,45 @@ def test_trainable_syncer_default(ray_start_2_cpus, temp_data_dirs):
     assert_file(False, tmp_target, os.path.join(checkpoint_dir, "checkpoint.data"))
 
 
+@pytest.mark.parametrize("num_retries", [None, 1, 2])
+def test_trainable_syncer_retry(ray_start_2_cpus, temp_data_dirs, num_retries):
+    """Check that Trainable.save() default syncing can retry"""
+    tmp_source, tmp_target = temp_data_dirs
+    num_retries = num_retries or 3
+
+    class FaultyCheckpoint(Checkpoint):
+        def to_uri(self, uri: str) -> str:
+            raise subprocess.CalledProcessError(-1, "dummy")
+
+    class TestTrainableRetry(TestTrainable):
+        _checkpoint_cls = FaultyCheckpoint
+
+        def _maybe_save_to_cloud(self, checkpoint_dir: str) -> bool:
+            from ray.tune.trainable.trainable import logger
+
+            output = []
+
+            def mock_error(x):
+                output.append(x)
+
+            with patch.object(logger, "error", mock_error), patch.dict(
+                os.environ,
+                {
+                    "TUNE_CHECKPOINT_CLOUD_RETRY_WAIT_TIME_S": "0",
+                    "TUNE_CHECKPOINT_CLOUD_RETRY_NUM": str(num_retries),
+                },
+            ):
+                ret = super()._maybe_save_to_cloud(checkpoint_dir)
+            assert f"after {num_retries}" in output[0]
+            return ret
+
+    trainable = ray.remote(TestTrainableRetry).remote(
+        remote_checkpoint_dir=f"file://{tmp_target}"
+    )
+
+    ray.get(trainable.save.remote())
```

Comment:
A question I have is how users would actually set the environment variables here. Since this is the primary "API" we should probably test this, e.g. using runtime env instead of in a patch.